### PR TITLE
new readline library

### PR DESCRIPTION
**Pull request by pdevine**
This is simplified version of the readline library which cuts out a lot of the complexity of the version that we were using. There's still a few things to add like "history" and getting the multi-line prompts working correctly, but most (many?) things should be more or less working, including:

* Each of the Ctrl-? chars (Ctrl-d, Ctrl-k, Ctrl-c, Ctrl-u, Ctrl-a, Ctrl-e, Ctrl-l, etc.)
* Line wrap with backspace/arrow keys
* Entering/exiting raw mode

Would love some feedback if people could try it out.

### CODE REVIEW

File: cmd/cmd.go
Comment (jmorganca) [inline cmd/cmd.go:496]:
```suggestion
```

Comment (pdevine) [reply cmd/cmd.go:496]: This is left over from the history stuff which will go back in soon.

File: readline/buffer.go
Comment (jmorganca) [inline readline/buffer.go:6]: Check out https://pkg.go.dev/container/list – might this be a good built-in alternative?

Comment (pdevine) [reply readline/buffer.go:6]: It would work, but it doesn't have a `Get()` function to get a node at a specific index. You have to keep track of that yourself.

Comment (jmorganca) [reply readline/buffer.go:6]: 👍

Comment (jmorganca) [reply readline/buffer.go:6]: Not important for merging, but would be curious if using `[]rune` could work– maybe with a custom type `type buf []rune` with some custom methods for the inserts/removes

Comment (pdevine) [reply readline/buffer.go:6]: I do use `[]rune` when inserting lines into the history (which is now added). This is similar to how the original golang readline works. For the individual characters it uses a single rune as nodes in the linked list.

Back to the original point, I could write a convenience function to replace `Get()` which just iterates from the head to correct index. I'm assuming that's probably what is going on with the gods implementation. `Remove()` could work somewhat similarly. We can look at adding that in the future to cut out the dependency.

File: cmd/cmd.go
Comment (pdevine) [reply cmd/cmd.go]: Ended up cutting this out and just sticking it directly inside of the new readline code.

**Comment by jmorganca** [comment]
🚀 

Small few things (optionally for later):

* alt+left and alt+right
* `tab` key seems to throw it out of wack
* history: when pressing down it to the "current message" would be awesome if it kept the buffer I had

File: readline/history.go
Comment (BruceMacD) [inline readline/history.go:42]: This looks like it could hit the dreaded "path does not exist" problem

**Review by BruceMacD — COMMENTED** [review]
Mostly worried about windows behavior, so far so good

File: readline/term.go
Comment (pdevine) [reply readline/term.go:9]: Yeah, it's copied from the original readline.

File: readline/term_bsd.go
Comment (pdevine) [reply readline/term_bsd.go:1]: With WSL it should work fine, but non-WSL it'll have issues. I think the non-WSL case can come in a follow-up PR; I do have part of a fix, but am having problems setting up a Windows dev environment.

File: readline/types.go
Comment (pdevine) [reply readline/types.go:68]: They conflict with the other constants. Golang doesn't like conflicting constants in the same constant block.

File: readline/types.go
Comment (pdevine) [reply readline/types.go:60]: I was lumping the terminal control stuff together in the same const block.

Comment (pdevine) [reply readline/types.go:64]: To be clear here, these terminal codes are all for controlling the screen, not for handling key presses.

File: cmd/cmd.go
Comment (BruceMacD) [inline cmd/cmd.go:601]: Should this be fixed before merge or will this happen in a follow up?

**Review by BruceMacD — COMMENTED** [review]
Validated in Warp. Seems pretty much good to go pending my question.

**Review by BruceMacD — APPROVED** [review]
tested and validated in terminal and warp

File: cmd/cmd.go
Comment (mxyng) [inline cmd/cmd.go:532]: nit:
```suggestion
		AltPlaceholder: `Use """ to end multi-line input`,
```

File: cmd/cmd.go
Comment (mxyng) [inline cmd/cmd.go:537]: debug print?

File: readline/buffer.go
Comment (mxyng) [inline readline/buffer.go:46]: nit: this doesn't really need fmt.Printf.

I think `Cursor(Up|Down|Left|Right)N` might be better suited as functions rather than constants.

e.g.

```go
var cursorUp = cursorUpN(1)

func cursorUpN(n int) string {
  return fmt.Sprintf("\033[%dA", n)
}
```

This becomes `fmt.Print(cursorUp + CursorBOL + cursorUpN(b.Width))`

File: readline/buffer.go
Comment (mxyng) [inline readline/buffer.go:48]:
```suggestion
			fmt.Print(CursorLeft)
```

File: readline/history.go
Comment (mxyng) [inline readline/history.go:47]: I don't think history should be world readable or writable. This file should at most be 600

Comment (mxyng) [inline readline/history.go:47]: Why not add `O_CREATE` here so it'll always exist?

File: readline/readline.go
Comment (mxyng) [inline readline/readline.go:232]: nit:

```suggestion
	r, ok := <-t.outchan
	if !ok {
		return rune(0)
	}
	return r
```

File: readline/buffer.go
Comment (pdevine) [reply readline/buffer.go]: I ended up implementing cursorUp/Down/Left/RightN.

**Review by jmorganca — APPROVED** [review]

**Comment by jikkuatwork** [comment]
Is this in `v0.1.5`? The model directory doesn't seem to change. 

I also tried to symlink the folder (to an NTFS external drive) but since the file names have colons, it failed. Any ideas?

**Pull request merged**
